src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs

```diff
@@ -25,13 +25,24 @@ internal class CodeActionResolveData
         /// </remarks>
         public string UniqueIdentifier { get; }
 
+        /// <summary>
+        /// Identifies the Code Action Provider which produced this Code Action.
+        /// </summary>
+        /// <remarks>
+        /// The unique identifier is currently set as:
+        /// name of top level code action provider + '|' + name of code action provider implementation
+        /// e.g. 'Add Await Code Action Provider | GetTitleWithConfigureAwait'
```

Comment:
Mirroring format of above [`UniqueIdentifier`](https://github.com/dotnet/roslyn/pull/48951/files#diff-620ffda894fe1518c66a7ac3d72474be7921240e332c3c36f196841d730bddb7L18-L26)

Reply:
`GetTitleWithConfigureAwait` is localized, which I believe would likely be problematic if Razor is trying to avoid localization issues.

I'm wondering if ProviderName can simply be made up of [PredefinedCodeFixProviderNames](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/PredefinedCodeFixProviderNames.cs,2c08df56a38dd211) and [PredefinedCodeRefactoringProviderNames](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/CodeRefactorings/PredefinedCodeRefactoringProviderNames.cs,86609280d63102d5)?
This would mean that you wouldn't be able to filter out specific code actions from a provider, but it would be simpler and would avoid localization issues.

Reply:
Note; we're doing [`nameof(GetTitleWithConfigureAwait)`](https://github.com/dotnet/roslyn/pull/48951/files#diff-28adfa208b8834665cb2fd3f0dca851f678f610d1e1506fb2f3cb9a7a73b0158R57) not using the result of the call.

It appears the [method name isn't localized](https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/CodeRefactorings/AddAwait/AddAwaitCodeRefactoringProvider.cs#L30-L31), however the [resource value is](https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/CSharpFeaturesResources.resx#L129-L134). In that case would we still run into localization issues? 

With regard to making things simpler, I agree removing it would reduce the complexity. Would you prefer that approach? 

My reasoning for specifying the implementation was, there may be cases where Razor may support one implementation of a provider, but not another. Granted, no specific example jumps out at me right now so this may be premature.

Reply:
Oh whoops, my bad, I totally missed that. 😅 Yeah, I don't think using `nameof` should have any localization issues.

In general, I do think the simpler approach would definitely be more ideal if that works for Razor. I don't think every code action provider will have something similar to `GetTitleWithConfigureAwait`, for example the Add Parameter code fix provider seems to have [different logic](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/AddParameter/AbstractAddParameterCodeFixProvider.cs,257) for displaying their code actions.

Reply:
Makes sense, thanks for the clarification! Will remove the implementation from the ProviderName.

Reply:
Also I think Manish may have previously mentioned something along these lines, but if Razor just uses `PredefinedCodeFixProviderNames` and `PredefinedCodeRefactoringProviderNames` as ProviderName, Roslyn could possibly expose these classes via external access so that Razor doesn't need to hardcode in any strings.

Reply:
Sounds good, thanks! What can we do to get the process started there for adding those provider names to external access? Appears we're all in agreement on the overall approach.

Reply:
Hmm, I'm not quite sure - @mavasani would you know how to proceed with moving PredefinedCodeRefactoringProviderNames and PredefinedCodeFixProviderNames to IVT?

Reply:
You will need to define helper types say, `RazorPredefinedCodeRefactoringProviderNames` and `RazorPredefinedCodeFixProviderNames` with constant string fields that forward to the internal types `PredefinedCodeRefactoringProviderNames` and `PredefinedCodeFixProviderNames` respectively. Something similar to https://github.com/dotnet/roslyn/blob/master/src/Tools/ExternalAccess/FSharp/Diagnostics/FSharpIDEDiagnosticIds.cs. I presume the new files will go https://github.com/dotnet/roslyn/tree/master/src/Tools/ExternalAccess/Razor?

Reply:
What's the difference in the code (refactoring/fix) providers in the `PredefinedCodeRefactoringProviderNames` and `PredefinedCodeFixProviderNames` and something like http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/GenerateConstructorFromMembers/AbstractGenerateConstructorFromMembersCodeRefactoringProvider.cs,39 where the `Add Null Checks` is a private field within a class?

Reply:
For Razor external access, can just the `PredefinedCodeRefactoringProviderNames` and `PredefinedCodeFixProviderNames` be exposed? I guess I'm not understanding why `AddNullChecksId` or `FeatureResources` might be necessary to expose. For example, for `AddNullChecksId`, is Razor able to simply use the overarching `PredefinedCodeRefactoringProviderNames.PredefinedCodeRefactoringProviderNames.GenerateConstructorFromMembers`?

Reply:
> For example, for AddNullChecksId, is Razor able to simply use the overarching PredefinedCodeRefactoringProviderNames.GenerateConstructorFromMembers

Ah, I wasn't aware that `AddNullChecks` would be a part of the `GenerateConstructorFromMembers`. [I tried annotating the associated abstract classes of `GenerateConstructorFromMembers`](https://github.com/dotnet/roslyn/pull/48951/commits/a16be50f92645feae7415105e8cd9666c1db776c#diff-25e13c2ac31208641a83ece9e79a7b1c7a470ebc25d8a843cb68acd8c6b6043bR28) but that didn't actually annotate the `Add Null Check` CodeAction.Data payload with the expected provider name. Is there a specific (inherited) code action I may have missed annotating?

Reply:
Ah, wait. The "Add null checks" feature resource you linked isn't actually a code action, I believe it's just part of the Generate constructor dialog (see number 4 image at https://docs.microsoft.com/en-us/visualstudio/ide/reference/generate-constructor?view=vs-2019).

When you say you want "Add null check" code action, are you possibly thinking of the action associated with the AbstractAddParameterCheckCodeRefactoringProvider? http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs,96